mdtk/downloaders.py

```diff
@@ -0,0 +1,233 @@
+"""
+Classes to download data from different source. Each class gets data from one
+source. The base data to get is midi. To contribute a new dataset, create a
+new class which extends DataDownloader, and write an accompanying test in
+./tests/test_downloads.py
+"""
+import os
+import sys
+import shutil
+import urllib
+import warnings
+import zipfile
+import glob
+from tqdm import tqdm
+
+
+
+USER_HOME = os.path.expanduser('~')
+DEFAULT_CACHE_PATH = os.path.join(USER_HOME, '.mdtk_cache')
+
+
+
+def download_file(source, dest, verbose=True, overwrite=None):
+        """Get a file from a url and save it locally"""
+        if verbose:
+            print(f'Downloading {source} to {dest}')
+        if os.path.exists(dest):
+            if overwrite is None:
+                warnings.warn(f'WARNING: {dest} already exists, not '
+                              'downloading', category=UserWarning)
+                return
+            if not overwrite:
+                raise OSError(f'{dest} already exists')
+        try:
+            urllib.request.urlretrieve(source, dest)
+        except urllib.error.HTTPError as e:
+            print(f'Url {source} does not exist', file=sys.stderr)
+            raise e
+
+
+def make_directory(path, overwrite=None, make_parents=True, verbose=True):
+        """Convenience function to create a directory and handle cases where
+        it already exists.
+        
+        Args
+        ----
+        path: str
+            The path of the directory to create
+        overwrite: boolean or None
+            If the path already exists, if overwrite is: True - delete the 
+            existing path; False - return error; None - leave the existing
+            path as it is and throw a warning    
+        make_parents: bool
```

Comment:
I don't think the make_parents flag is necessary. I would default to os.makedirs. Is the use case of this is to catch typos (i.e. cases where the parent SHOULD exist, but doesn't)? That can be fixed with a simple cut+paste or mv command by the user.

Reply:
That's fair - it's just to force the user to be explicit. But since I've set default to `True` anyway. I'll remove, we can reintroduce if required I guess.